outlines/models/hf_transformers.py

```diff
@@ -421,3 +421,107 @@ def create_type_constraint(
         raise NotImplementedError(f"Cannot restrict the generation to type {type}")
 
     return type_to_mask[type](tokenizer, prompt_tokens)
+
+
+def HuggingFaceEmbeddings(model_name: str):
+    """Create a function that will download and run a HuggingFace embedding locally.
+
+    You should have the `transformers` package installed. Available models are listed
+    on the `HuggingFace documentation <https://huggingface.co/sentence-transformers`_.
+
+    Note: The first time this is run it might take 20-30 seconds to download the model weights.
+
+    Parameters
+    ----------
+    model_name: str
+        The model name as listed in the HuggingFace website.
+
+    Returns
+    -------
+    A function that will call run the HuggingFace embedding with the given parameters when
+    passed a prompt. It will use a GPU if there is one available.
+
+    """
+
+    def mean_pooling(model_output, attention_mask):
+        """Pools together the outputs from a sentence embedding model to generate a vector embedding."""
+        import torch
+
+        token_embeddings = model_output[0]
+        input_mask_expanded = (
+            attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
+        )
+        return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(
+            input_mask_expanded.sum(1), min=1e-9
+        )
+
+    def get_embedding(
+        sentences: Union[List[str], str], *, batch_size: Optional[int] = None
+    ):
+        """Generates embeddings from mean-pooled sentence embeddings from HuggingFace sentence transformers.
+
+        Parameters
+        ----------
+        sentences: List[str] | str
+            The strings to be embedded
+        batch_size: Optional[int]
+        The batch size. If it is not provided, or if a negative value is given, the embeddings will be run as a single batch.
+
+
+        Returns
+        -------
+        A function that will call run the HuggingFace embedding with the given parameters when
+        passed a prompt. It will use a GPU if there is one available.
+
+        """
+        import torch
+        from transformers import AutoModel, AutoTokenizer
```

Comment:
Is there any kind of check on `model_name` that we can enforce? For text generation this is done on the `transformers` size by calling `AutoModelForCausalLM`. Is there a similar model class in `transformers` that checks that the model is appropriate for the given task?

Maybe even something as simple as making sure that the model name starts with `sentence-transfomers/`.